### PR TITLE
Refactor modified components

### PR DESCRIPTION
**Pull request by PaulFarault**
#### Which issue(s) this PR fixes

Part of #348 

#### Additional comments

Refactor `ServiceVariables.component_modified`:

- Rewrite `ServiceVariables.component_modified` to :
  
  ```diff
  - ServiceVariables.components_modified(self, dag: Dag, version: str) -> List[Operation]
  + ServiceVariables.modified_components(self, collections: Collections, version: str) -> Set[Tuple[str, str]]
  ```

- Create `ClusterVariables.modified_component` to reproduce the result of `ServiceVariables.component_modified` on the whole configuration.
- Create methods to check is a component has been modified:
  - `ServiceVariables.is_service_component_modified_from_version(service_component: ServiceComponentName, version: str) -> bool`
  - `Repository.is_file_modified(commit: str, file: Union[str, os.PathLike]) -> bool`
- Update the consumer functions accordingly (`tdp plan reconfigure` command in particular).
- Create `Collections.get_service_component(service: str) -> List[str]` to easily recover components for a single service.
- Fix tests.

#### Agreements

- [X] I hereby declare this contribution to be licensed under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0).
- [X] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

**Comment by PaulFarault** [comment]
Todo: Transform the `_VariablesFileInfos` class to a `ServiceComponent` class as it is never clear how to manipulate components (as `service_component` string, as `(service, component)` tuple or only `component` string).

**Comment by PaulFarault** [comment]
Interfaces are better now with the creation of `ServiceComponentName`. I'm not entirely satisfied with `ServiceVariables.get_modified_components_names` but I didn't find a better implementation yet.

If you're ok with this version @rpignolet, I'll squash commits, lint code and add tests.

File: tdp/core/deployment/deployment_plan.py [hunk 1]
```diff
@@ -134,41 +138,48 @@ def from_operations(operations: List[Operation]) -> "DeploymentPlan":
...
+        operations = set()
+        modified_components = cluster_variables.get_modified_components_names(
+            deployed_component_version_logs,
+        )
+        for modified_component in modified_components:
+            logger.debug(
+                f"Component {modified_component.component_name} of service {modified_component.service_name} has been modified"
+            )
+            is_service = modified_component.component_name is None
+            if is_service:
```
Comment (rpignolet) [inline tdp/core/deployment/deployment_plan.py:165]: You can use the `is_service` method no ?
```suggestion
            if modified_component.is_service():
```

File: tdp/core/service_component_name.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        """
+        [service_name, *component_name] = full_name.split("_", 1)
+        component_name = component_name[0] if component_name else None
+        return ServiceComponentName(service_name, component_name)
+
+    def __repr__(self) -> str:
+        return f"ServiceComponentName({self.service_name}, {self.component_name})"
+
+    def __str__(self) -> str:
+        return self.full_name
```
Comment (rpignolet) [inline tdp/core/service_component_name.py:44]: Missing a new line ?

**Review by rpignolet — APPROVED** [review]

**Comment by sergkudinov** [comment]
I rebased this to the branch of #371, it has multiple broken tests:

```
tdp/core/deployment/test_deployment_plan.py::test_deployment_plan_reconfigure FAILED
tdp/core/deployment/test_deployment_runner.py::test_deployment_is_reconfigured FAILED
tdp/core/deployment/test_deployment_runner.py::test_deployment_reconfigure_is_resumed FAILED
```

**Comment by PaulFarault** [comment]
There was a typo, it should be ok now!

**Review by sergkudinov — APPROVED** [review]
LGTM

**Pull request merged**